source/common/upstream/subset_lb.h

```diff
@@ -111,10 +119,19 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
   typedef std::vector<std::pair<std::string, ProtobufWkt::Value>> SubsetMetadata;
 
   class LbSubsetEntry;
+  struct SubsetSelectorMap;
+
   typedef std::shared_ptr<LbSubsetEntry> LbSubsetEntryPtr;
+  typedef std::shared_ptr<SubsetSelectorMap> SubsetSelectorMapPtr;
   typedef std::unordered_map<HashedValue, LbSubsetEntryPtr> ValueSubsetMap;
   typedef std::unordered_map<std::string, ValueSubsetMap> LbSubsetMap;
 
+  struct SubsetSelectorMap {
```

Comment:
I think it is fine as it is. However you could likely simplify this. The subset fallback policy is fully identified by the set of selector keys for this subset. You can concatenate the sorted selector keys into one string and use it as a key for finding the fallback policy. In this way you do not need to build a trie and can just use std::unordered_set<std::string, envoy::api::v2::Cluster::LbSubsetConfig::LbSubsetSelectorFallbackPolicy>

Reply:
@yanavlasov agree that that way code would be simplified. Was thinking about tradeoff though, if to switch to plain map, on each lookupHosts call LB will have to perform concatenation of keys from match criteria to look it up in map. While first approach brings code complexity but is kind of precomputing things on LB object creation so that when lookupHosts called we just iterate over each match criteria key and look it up recursively in a trie. If it's not a big deal of match criteria keys concatenation (from memory usage perspective), then having plain map seems more beneficial.